### PR TITLE
Dict merge(!) with combiner function

### PR DESCRIPTION
**Pull request by fredrikekre**
Implements part A of https://github.com/JuliaLang/julia/issues/20699

Did not find any tests for `merge(!)` without combiner function so I added that.

A bit of code duplication though. Something like this:

```diff
diff --git a/base/associative.jl b/base/associative.jl
-function merge!(d::Associative, others::Associative...)
-    for other in others
-        for (k,v) in other
-            d[k] = v
-        end
-    end
-    return d
-end
+merge!(d::Associative, others::Associative...) = merge!((i, j)-> j, d, others...)
```

could be applied but would result in a slower `merge!`. Thoughts?

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Let's leave the bit of duplication here.

**Review by StefanKarpinski — APPROVED** [review]

File: base/associative.jl [hunk 1]
```diff
@@ -215,13 +253,45 @@ Dict{String,Float64} with 3 entries:
...
+  "bar" => 4753.0
+  "baz" => 17.0
+  "foo" => 0.0
+```
+"""
+merge(combine::Function, d::Associative, others::Associative...) =
+    merge!(combine, emptymergedict(d, others...), d, others...)
+
+function emptymergedict(d::Associative, others::Associative...)
     K, V = keytype(d), valtype(d)
```
Comment (fredrikekre) [inline base/associative.jl]: Is there an example on how to do this somewhere?

(And just for my own understanding; the behaviour of the implementation in this PR is the same as the previous one, right?)

Comment (stevengj) [inline base/associative.jl]: See, for example, how the `promote_type` function itself is implemented.

Comment (stevengj) [inline base/associative.jl]: something like
```jl
promoteKV(K,V) = (K,V)
promoteKV(K,V, d, ds...) = promoteKV((promote_type(K,keytype(d)), promote_type(V,valtype(d))), ds...)
```
and then call `K,V = promoteKV(keytype(d), valtype(d), others...)`

Comment (KristofferC) [inline base/associative.jl]: You usually need some forced `@inlining` as well, or?

Comment (stevengj) [inline base/associative.jl]: Usually not, I thought.

Comment (fredrikekre) [inline base/associative.jl]: Alright. How can I test that it is type-inferrable? Benchmarks?

Comment (KristofferC) [inline base/associative.jl]: `@inferred`?

Comment (fredrikekre) [inline base/associative.jl]: I was sure I tried that and that it didn't complain. But it does.

```julia
julia> d1 = Dict("A" => Int32(1), "B" => Int32(2));

julia> d2 = Dict("A" => 1, "B" => 2);

julia> @inferred Base.emptymergedict(d1, d2)
ERROR: return type Dict{String,Int64} does not match inferred return type Dict{String,_} where _
Stacktrace:
 [1] error(::String) at ./error.jl:21
```

Comment (stevengj) [inline base/associative.jl]: @fredrikekre is that with the old version or with the version I suggested?

Comment (fredrikekre) [inline base/associative.jl]: Whats currently in this PR. Trying out your suggestion right now.

Comment (fredrikekre) [inline base/associative.jl]: This implementation:

```julia
promoteKV(K,V) = (K,V)
promoteKV(K,V, d, ds...) = promoteKV(promote_type(K, keytype(d)), promote_type(V, valtype(d)), ds...)
function Base.emptymergedict(d::Associative, others::Associative...)
    K, V = promoteKV(keytype(d), valtype(d), others...)
    Dict{K,V}()
end
```

results in:
```julia
julia> d1 = Dict(Int32(1) => 1, Int32(2) => 2);

julia> d2 = Dict(1 => 1.0, 2 => 2.0);

julia> @inferred Base.emptymergedict(d1, d2)
ERROR: return type Dict{Int64,Float64} does not match inferred return type Dict{_,_} where _ where _
Stacktrace:
 [1] error(::String) at ./error.jl:21
```

But I tried this:
```julia
promoteK(K) = K
promoteV(V) = V
promoteK(K, d, ds...) = promoteK(promote_type(K, keytype(d)), ds...)
promoteV(V, d, ds...) = promoteV(promote_type(V, valtype(d)), ds...)
function Base.emptymergedict(d::Associative, others::Associative...)
    K = promoteK(keytype(d), others...)
    V = promoteV(valtype(d), others...)
    Dict{K,V}()
end
```
which seems to work:

```julia
julia> d1 = Dict(Int32(1) => 1, Int32(2) => 2);

julia> d2 = Dict(1 => 1.0, 2 => 2.0);

julia> @inferred Base.emptymergedict(d1, d2)
Dict{Int64,Float64} with 0 entries
```

Should I update to that version? Should `promoteK/V` be inlined?

Comment (stevengj) [inline base/associative.jl]: the promoteK/V version looks ok as-is; I wonder why the tuple version didn't work?

**Review by StefanKarpinski — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
Technically past feature freeze, but there's no alpha out yet, so I think we can sneak this in.

**Pull request merged**